panel-plugin/i3w-multi-monitor-utils.c

```diff
@@ -0,0 +1,119 @@
+#include "i3w-multi-monitor-utils.h"
+
+/**
+ * find_crtc:
+ * @dpy: The X Display
+ * @res: The XRandR screen resources
+ * @xid: The X Id of the crtc
+ *
+ * Find a crtc's information given its xid
+ *
+ * Returns: The crtc's information struct
+ */
+XRRCrtcInfo*
+find_crtc (Display* dpy, XRRScreenResources* res, RRCrtc xid) {
+    XRRCrtcInfo* found = NULL;
+    int c;
+    for (c = 0; c < res->ncrtc && !found; ++c) {
+        if (xid == res->crtcs[c]) {
+            found = XRRGetCrtcInfo(dpy, res, res->crtcs[c]);
+        }
+    }
+    return found;
+}
+
+
+/**
+ * is_connected:
+ * @plugin: the xfce plugin object
+ *
+ * Does c mean connected display?
+ *
+ * Returns: True iff c indicated connected display
+ */
+gboolean
+is_connected (Connection c) {
+    return c == RR_Connected;
+}
+
+
+/**
+ * get_outputs:
+ *
+ * Gets the output information from XRandR. For each monitor, the
+ * output name, screen position and current resolution.
+ *
+ * Returns: The outputs structure. Must be freed with free_outputs.
+ */
+i3_workspaces_outputs_t
+get_outputs() {
+    // Get Xlib information
+    Display* dpy = XOpenDisplay (NULL); 
+	int screen = DefaultScreen (dpy);
+    Window root = RootWindow (dpy, screen);
+	XRRScreenResources* res = XRRGetScreenResourcesCurrent (dpy, root);
+    int num_outputs = res->noutput;
+
+    // Allocate output
+    i3_workspaces_outputs_t outputs;
+    outputs.outputs = malloc(sizeof(i3_workspaces_output_t)*num_outputs);
+
+    // NOTE: We will only store connected outputs, even though we allocated space
+    // for all outputs. This is better than having to count the outputs first and 
+    // asking for space later.
+    int num_connected_outputs = 0; 
+    int o;
+    for (o = 0; o < num_outputs; ++o) {
+        XRROutputInfo *output_info = XRRGetOutputInfo(dpy, res, res->outputs[o]);
+        int connected = is_connected(output_info->connection);
+
+        if (connected) {
+            XRRCrtcInfo* info = find_crtc(dpy, res, output_info->crtc);
+            if (info) {
+                i3_workspaces_output_t* output = &outputs.outputs[num_connected_outputs];
+
+                output->name = malloc(strlen(output_info->name)*sizeof(char));
```

Comment:
You need `strlen(output_info->name) + 1` as `strlen()` doesn't include the null terminator.